Refactor Modifiers

TimothyWillard opened the pull request:
### Describe your changes.

- [X] Add pydantic representations for distributions with a sample method for usage here. Ideally would also be used in likelihood, but should reserve this for a follow up issue+PR.
- [x] Add Gamma, Beta, and Weibull distributions, #573.
- [x] Add unit tests for existing `gempyor.NPI` functionality, to build a baseline of existing behavior as needed.
- **nb: moving from "reduce" to "apply modifiers"**
- [x] Add `apply` abstract method to `ModifierABC` taking arguments `parameter_name`, `parameter`, `method` and using `SinglePeriodModifier` as an example implement `reduce`. **nb: see helpers.py for ex.** Should drop into: https://github.com/HopkinsIDD/flepiMoP/blob/00a262d7ee9307dbf7956133bc28d2aa8bbd13d0/flepimop/gempyor_pkg/src/gempyor/parameters.py#L309-L336.
- [x] Add `apply_dataframe` method to `ModifierABC` that is a wrapper around `apply`.
- [x] Add `ModifiersCollection` using `pydantic.BaseModel` to represent the 'outcome_modifiers' and 'seir_modifiers' sections in config files. Pretty minimal, `scenarios: list[str]` and `modifiers: list[Modifier]`.
- [x] Refactor `ModifiersCollection` to have a `scenarios_map` attribute (needs better name or maybe some rethinking, ideally would be called `scenarios` but that attribute is already occupied) that is `dict[str, list[str]]` to represent "stacked modifiers". Requires adding a model validator to `ModifiersCollection` to rewrite `StackedModifier`s to this to maintain backwards compatibility.
- [x] Add `reduce` (**now .apply()**) method to `ModifiersCollection` that implements similar logic to the current `StackedModifier` that takes a `scenario` and modifies multiple parameters at once.
- [ ] Implement `MultiPeriodModifier`. (**ie, incorporate MultiPeriodModifier into PeriodicModifier**)
- [ ] Address #490 if not already addressed by stricter validation (might require following up for details/cmds).
- [ ] Integrate `ModifiersCollection` into `gempyor.seir/outcomes`.
- [ ] Explore possibilities for https://github.com/HopkinsIDD/flepiMoP/pull/570#discussion_r2144395930. Doable with some meta-programming, but loose the ability to validate inputs at configuration parse time.
- [ ] Ensure renaming from "reduce" to "apply modifiers"
- [ ] `git rm` the NPI subpackage.


### Does this pull request make any user interface changes? If so please describe.

The user interface changes are...

Those are reflected in updates to the documentation in ...

### What does your pull request address? Tag relevant issues.

This pull request addresses: #409.

pearsonca reviewed (commented):
Just some thoughts on the work in progress

emprzy committed:
Fix error in test function naming

There was a doubly-named test in `test_npis.py` that also contained an outdated loc .

emprzy committed:
Finish `gempyor.NPI` unit test library.

emprzy committed:
Add `apply` as abstract method to `ModifierABC`

Implement this method in `_single_period::SinglePeriodModifier` class.

emprzy committed:
Add `apply` as abstract method to `ModifierABC`

Implement this method in `_single_period::SinglePeriodModifier` class.

emprzy committed:
Add `.apply_dataframe()` method to `ModifierABC`

`.apply_dataframe()` is a wrapper around `.apply()`

emprzy committed:
Add `.apply_dataframe()` method to `ModifierABC`

`.apply_dataframe()` is a wrapper around `.apply()`

jcblemai commented:
I could not really read the diff which is huge, and I don't follow much flepi but my 2-cent is that I would ditch the recursive autocall modifier thing, which makes anything like smart inference hard. I would load them all at once from config. 

config breaking change I think are necessary:
* Every modifier is case of multi-timereduce so I would just keep this class to ease up the syntax. Probably something can be made to keep config for functioning
* I would probably ditched the stacked-modifier (not now perhaps) because it is confusing. But this is up to debate

emprzy committed:
Add `ModifiersCollection`

Additionally, rename `SinglePeriodModifier` to `PeriodicModifier` (now represents both single and mulitperiod)

emprzy committed:
Add `ModifiersCollection`

Additionally, rename `SinglePeriodModifier` to `PeriodicModifier` (now represents both single and mulitperiod)

emprzy committed:
Add to `ModifiersCollection`

emprzy committed:
Add to `ModifiersCollection`

emprzy committed:
Hide `_scenarios` as an internal attribute

emprzy committed:
Hide `_scenarios` as an internal attribute
